### PR TITLE
feat(googleapps): add support for 'extra number' in device push challenges

### PR DESCRIPTION
**Pull request by nickvollmar**
This PR addresses issues https://github.com/Versent/saml2aws/issues/477 and https://github.com/Versent/saml2aws/issues/728.

In some cases, Google's device push challenges come with an 'extra number', which the user must enter on their phone for the prompt to succeed. Personally, I get an 'extra number' whenever I attempt to sign in from a new WiFi network. (To reproduce, you can use a VPN and log in from a country that you've never been to.)

#### Example output with 'extra number' challenge
```
Using IdP Account default to access GoogleApps https://accounts.google.com/o/saml2/initsso?...
Authenticating as nick@example.com ...
Check your phone and tap 'Yes' on the prompt, then tap the number:
	46
Then press ENTER to continue.

Selected role: arn:aws:iam::123456789012:role/foobar
Requesting AWS credentials using SAML assertion.
Logged in as: arn:aws:sts::123456789012:assumed-role/foobar/nick@example.com
```

#### Example output without extra number
```
Using IdP Account default to access GoogleApps https://accounts.google.com/o/saml2/initsso?...
Authenticating as nick@example.com ...
Check your phone and tap 'Yes' on the prompt. Then press ENTER to continue.

Selected role: arn:aws:iam::123456789012:role/foobar
Requesting AWS credentials using SAML assertion.
Logged in as: arn:aws:sts::123456789012:assumed-role/foobar/nick@example.com
```

The jsname attribute is being used elsewhere in this file (`func extractKeyHandles`) so I hope it's safe to use here too.

### CODE REVIEW

**Comment by alsaghir** [comment]
Is there any idea when to merge this ?

**Comment by miguelaferreira** [comment]
I would also very much like this feature to be merged and releases. My team has had to disable google prompt altogether since we reported #728.

**Comment by TWRogers** [comment]
Seconded

**Comment by nickvollmar** [comment]
Hello @wolfeidau, I'm sorry to bother you –– but I notice you've merged some PRs lately!

May I ask for review on this change?

I will happily sign any contributor/copyright agreement if Versent needs that.

**Comment by martin-shippit** [comment]
I've tried this PR and looks like it solves the issue. Nice work. My only concern would be the lack of tests. Perhaps it's not feasible :thinking:

**Review by martin-shippit — APPROVED** [review]

**Comment by nickvollmar** [comment]
Hello @martin-shippit, sorry for the delay. I have added a test and merged the latest changes from upstream. Could you re-approve and perhaps merge?

**Comment by martin-shippit** [comment]
Beautiful. I don't have merge privilege but this looks good to me. I've been using this version for a while and it works well.

@wolfeidau could you review and merge if it looks good to you?

**Comment by jmctune** [comment]
@wolfeidau :wave: Several users in my org get stuck with this exact use case and have to either guess the correct number or hope Google presents them with a different challenge. Could this be reviewed / merged?

**Comment by nickvollmar** [comment]
Thank you @mapkon! I believe I have addressed the linter errors, could you please re-run the GitHub action?

**Comment by mapkon** [comment]
> Thank you @mapkon! I believe I have addressed the linter errors, could you please re-run the GitHub action?

Superb! I will take a look - please watch out for the code coverage, incase it regresses.

File: pkg/provider/googleapps/googleapps.go [hunk 1]
```diff
@@ -737,3 +744,11 @@ func isAppId(val string) string {
 	}
 	return appId
 }
+
+func extractDevicePushExtraNumber(doc *goquery.Document) string {
+	extraNumber := ""
+	doc.Find("div[jsname=feLNVc]").Each(func(_ int, s *goquery.Selection) {
```
Comment (mapkon) [inline pkg/provider/googleapps/googleapps.go:750]: I am wondering if it is possible for this to change? It looks like a generated name, which might  be different from page to page?

Comment (nickvollmar) [reply pkg/provider/googleapps/googleapps.go:750]: I agree that this is a generated name, and therefore it's less than ideal to use. But it seems to me that this is the best we can do. 

I checked in an anonymized version of the HTML at `pkg/provider/googleapps/example/challenge-extra-number.html` –– as you can see, if we don't want to use generated names, we have to couple ourselves to the HTML structure, which is not ideal either.

A couple of mitigating circumstances that make me feel better about the usage of jsname:
1. Elsewhere in this file we have `func extractKeyHandles` that relies on a jsname as well.
2. [@martin-shippit's comment on Aug 31 '22](https://github.com/Versent/saml2aws/pull/777#issuecomment-1233520239) indicates that this logic worked for him too, so his page must have had the same jsname.

Comment (mapkon) [reply pkg/provider/googleapps/googleapps.go:750]: OK, thanks for the explanation.

**Review by mapkon — APPROVED** [review]

**Pull request merged**